### ISSUE TITLE
Allow modification of published variants that are added to components by plugins

### ISSUE THREAD

**Issue by jjohannes** [gradle/gradle#10428]
Plugins, for example the Java plugins, can add addition variants to components for publishing automatically (e.g. for _test fixtures_ or other _features_). Sometimes, it might be necessary to adjust these variants before publishing. For example, if publishing test fixtures is not desired.

The proposed solution is to add a `withVariantsFromConfiguration(Configuration, Action<ConfigurationVariantDetails>)` to the [AdhocComponentWithVariants](https://github.com/gradle/gradle/blob/2f4648f5479d8780741584fd6bc940d48d1b4870/subprojects/core-api/src/main/java/org/gradle/api/component/AdhocComponentWithVariants.java) interface in parallel to the `addVariantsFromConfiguration()` method.

This would allow you to do for example:

```
val javaComponent = components["java"] as AdhocComponentWithVariants
javaComponent.withVariantsFromConfiguration(configurations["testFixturesApiElements"]) {
    skip()
}
javaComponent.withVariantsFromConfiguration(configurations["testFixturesRuntimeElements"]) {
    skip()
}
```

**Comment by ljacomet** [comment]
I think it makes sense but I wonder if we really need two different methods for this on `AdhocComponentWithVariants`.
We could have a single method acting as an optional add + register configuration action, allowing multiple configuration actions to be registered.

If not, this method should throw when the configuration has not been registered previously.
This would allow a user that added the `java-test-fixtures` plugin and customized publication to be notified the customization is no longer needed if he removes the plugin.

**Comment by jjohannes** [comment]
We decided to add the additional method, but make it fail if the variant it adjusts does not exist.

Also:
- Document the feature in _publishing docs_
- Document how to use it to disable test fixture publishing in _text fixture docs_
